### ISSUE TITLE
Make it easier to create custom Layers

### ISSUE THREAD

**Issue by bauerdavid** [napari/napari#3253]
## 🚀 Feature

Allowing developers to easily create custom layers, and (if this is already possible) create a tutorial for it (what to implement, how to use callbacks, etc.)

## Motivation
I am working on a module to allow the user to select parts of the 3D (nD) data, so if they visualize it in 3D, then they do not have to worry about filling the memory. For this I implemented a custom BoundingBox layer, but actually it is kind of hacky, and also I could not find out how can I create a plugin from it. So it would be quite nice if there was a nice and clear way of implementing any custom layer.

## Pitch
Maybe this is already possible, but I didn't find any tutorial on how to do this. Either way, it would be nice to have a clear and easy-to-follow description on how to implement custom layers.
It would also be good if the developer could include it in napari in a way that users can use them like built-in layers (e.g. a 'create XY layer' button appears in the GUI)
Generalize the layer attributes widgets so it is not needed to be implemented manually. For example, there is a list of modes which the user can select, including its name, icon. And this also stands for the custom attributes of the layer. What I'm thinking is that this should be similar in some sense to the `@magicgui` decorator (the needed widgets are inferred from the function attributes).
I'm not sure what would be the best way to also include the handling of mouse events. Right now I copied everything in connection to the Shapes layer (layer implementation, mouse callbacks, vispy visualization etc.),modified as needed and keep them together in the same directory, but it is quite ugly.

## Alternatives
I was also thinking about creating a module and simply leave the files used for the layer there, but I'm not sure if this is the way to go (for example, I can imagine that I want to use the same layer in multiple images). I think the best way would be to create a similar API for creating custom layers as for modules (e.g. `@napari_hook_implementation` decorator).

## Additional context
So actually what I'm trying to do (and mostly I managed to) implement the needed tools to create (n-dimensional) bounding boxes around multiple objects in the opened file, list each one of them, than allow the user to annotate them, one by one, hence making it possible to visualize the selected object in 3D without having to load the whole data into memory. I also created a widget which shows different projections of the selected object, also displaying the annotations. I attached a screenshot of the current state (yup, the annotator cursor was drawn in post, as it does not appear on the screenshot).
In some way I would like to make it available publicly, if you have any suggestions on this, I'm open to possibilities.

![08_annotate_w_pointer](https://user-images.githubusercontent.com/36735863/130972301-9ccd9b3a-8424-4757-9c5c-1ce4620d5255.png)

**Comment by sofroniewn** [comment]
Wow @bauerdavid I love what you're doing here! nD bounding boxes is something that we've actually wanted to add support for in napari too.

@jni has been quite keen on the idea making it easier to add custom layers, particularly as a way for us to grow functionality before deciding whether to incorporate something into the napari core or not.

One approach might be for us to add a `@napari_experimental_provide_layer` hook specification so plugins could provide custom layers. @tlambert03 and @nclack are currently leading some improvements to the plugin interface, so I'd be curious to get their take before suggesting that we press on with that approach.

I'm not sure if you've joined the [napari zulip](https://napari.zulipchat.com/login/) yet but we can also follow up with discussion there too or at one of our [community meetings](https://napari.org/community/team.html#meeting-schedule).

Looking forward to discussing more!!

**Comment by bauerdavid** [comment]
Yeah, the hook specification might be a great idea.

I didn't join zulip until now, but I already fixed this problem. Neither was I aware of the community meetings, but I will try to join soon. Are other (working group) meetings public? I don't think I can provide any input, just I would be interested in listening to the ongoing work (and maybe I would understand some concepts better).

Keep up the good work!

**Comment by jni** [comment]
Yup, working group meetings are also open. They should appear in the calendar as they are scheduled. Looking forward to seeing you around @bauerdavid!

And yes I do like the idea of allowing custom layers but right now that's not a thing and it's definitely not trivial, as you would need to provide a layer class a VispyLayer (contains all the stuff for drawing with OpenGL), and some controls. I would say we should probably wait to improve our slicing model before providing a custom layer plugin API. But that is high priority!

**Comment by bauerdavid** [comment]
Thank you very much!

The bounding box layer still needs some bugfix and minor changes, but I'm working on it. Are you interested in taking a look at it?

**Comment by sofroniewn** [comment]
> The bounding box layer still needs some bugfix and minor changes, but I'm working on it. Are you interested in taking a look at it?

Yes for sure! If you add a link I'll be happy to

**Comment by bauerdavid** [comment]
Okay, I will clean up some stuff, and send it soon. Thanks!

**Comment by bauerdavid** [comment]
@sofroniewn I invited you to the repo, feel free to try the code. I would be glad to get your opinions, suggestions, etc.
